eclair-core/src/main/scala/fr/acinq/eclair/router/Validation.scala

```diff
@@ -414,51 +414,55 @@ object Validation {
-        log.info(s"channel shortChannelId=$realShortChannelId is back from the dead! requesting announcements about this channel")
-        sendDecision(origins, GossipDecision.RelatedChannelPruned(u))
-        db.removeFromPruned(realShortChannelId)
-        // peerConnection_opt will contain a valid peerConnection only when we're handling an update that we received from a peer, not
-        // when we're sending updates to ourselves
-        origins.head match {
-          case RemoteGossip(peerConnection, remoteNodeId) =>
-            val query = QueryShortChannelIds(u.chainHash, EncodedShortChannelIds(routerConf.encodingType, List(realShortChannelId)), TlvStream.empty)
-            d.sync.get(remoteNodeId) match {
-              case Some(sync) if sync.started =>
-                // we already have a pending request to that node, let's add this channel to the list and we'll get it later
-                d.copy(sync = d.sync + (remoteNodeId -> sync.copy(remainingQueries = sync.remainingQueries :+ query, totalQueries = sync.totalQueries + 1)))
-              case _ =>
-                // otherwise we send the query right away
-                peerConnection ! query
-                d.copy(sync = d.sync + (remoteNodeId -> Syncing(remainingQueries = Nil, totalQueries = 1)))
-            }
-          case _ =>
-            // we don't know which node this update came from (maybe it was stashed and the channel got pruned in the meantime or some other corner case).
-            // or we don't have a peerConnection to send our query to.
-            // anyway, that's not really a big deal because we have removed the channel from the pruned db so next time it shows up we will revalidate it
-            d
+          db.getChannel(RealShortChannelId(u.shortChannelId.toLong)) match {
+            case Some(pc) =>
+              // The channel was pruned (one of the two channel updates was older than 2 weeks).
+              val pc1 = pc.applyChannelUpdate(update)
+              if (!Announcements.checkSig(u, pc.getNodeIdSameSideAs(u))) {
+                log.warning("bad signature for announcement shortChannelId={} {}", u.shortChannelId, u)
+                sendDecision(origins, GossipDecision.InvalidSignature(u))
+                d
+              } else if (pc1.isStale(currentBlockHeight)) {
+                log.debug("channel shortChannelId={} is still pruned", u.shortChannelId)
+                sendDecision(origins, GossipDecision.RelatedChannelPruned(u))
+                db.updateChannel(u)
+                d
+              } else {
+                log.info("channel shortChannelId={} is back from the dead", u.shortChannelId)
+                sendDecision(origins, GossipDecision.Accepted(u))
+                val latestUpdates = pc1.update_1_opt ++ pc1.update_2_opt
+                ctx.system.eventStream.publish(ChannelUpdatesReceived(latestUpdates))
+                db.updateChannel(u)
```

Comment:
Should you have instead called `addPublicChannel` (like we do for local channels that graduate from private to public) and `handleChannelUpdate` (like we do in `handleChannelValidationResponse` for stashed updates)? It guarantees that we do the exact same processing for announcements whatever the context (first seen or pruned). Btw, the behavior is changing compared to before: we may send a `ChannelUpdatesReceived` without a `ChannelDiscovered` depending on the pruning time and the node restart time.

Looking deeper into it, at first I thought that adding back the channel manually as opposed to calling `addPublicChannel` made us miss putting back the `WatchExternalChannelSpent`, but... it's already there because we watch _all_ (live and stale) channels on restart. _However_, the `handleChannelSpent` only checks the live channels map so there is indeed a bug (and probably a missing test).

There is another issue: not emitting `ChannelDiscovered` will not work in a clustered setup, because front routers rely on this event to find out about new channels and they will never be made aware of unpruned channels coming back to life.

Reply:
Partially done in https://github.com/ACINQ/eclair/pull/2456/commits/e048636104d35d8e12b5b01a94e20886d24dbcad for the channel updates.

I didn't call `addPublicChannel` though, as we don't want to revalidate the on-chain outpoint, that would be another DoS vector wouldn't it? Also we can't be in a private -> public transition if a channel has been pruned, it must have been public before. So it feels like the only part from `addPublicChannel` that we actually need is the `ChannelsDiscovered` event, so I imported only that, but let me know if you think we should refactor `addPublicChannel` to be able to call it here.

Reply:
You're right! But it does need a lot of tweaks to actually be used in our context:

- we don't want to waste a call to `db.addChannel`
- we don't want to rebroadcast the channel announcement to our peers
- there's no need to put the watch, it was already put at start-up
- all the private -> public code is unnecessary

Overall it feels like we only want a very small subset of what `addPublicChannel` does, that's why I'm a bit reluctant, it feels harder to reason about than copy-pasting the useful code...I tried refactoring it to extract the useful part, but it makes it actually even harder to read and check for correctness :/

Reply:
Fair enough :cowboy_hat_face: